### ISSUE TITLE
'?' failing to bring up hotkey menu on https://reviewable.io/reviews

### ISSUE THREAD

**Issue by ryan-williams** [Reviewable/Reviewable#56]
At https://reviewable.io/reviews, I type "?", or various modifiers + "?", and nothing seems to show me available hotkeys. I thought it would, per #12.

A UI element that allows viewing them would also be useful.

**Comment by pkaminski** [comment]
Currently hotkeys are only available on (and hence only show the cheatsheet on) individual review pages.  What shortcuts are you looking for on the reviews dashboard page?

Also, what do you mean by "UI element"?

**Comment by ryan-williams** [comment]
my intuition is that the shortcut to show me shortcuts should always be available. more shortcuts on that page would also be nice, like a j/k to select a review, enter to open, etc.

by UI element, I mean there could be a DOM element that is visible to the user, maybe in page chrome, that allows them to view keyboard shortcuts. 

for example, Asana teases some shortcuts and links to "more" at the bottom of most/all pages:

![](http://f.cl.ly/items/0n1x290M2d1w34061s3Y/Screen%20Shot%202015-02-27%20at%201.51.58%20PM.png)

tbh that's probably more than I need, really something functionally equivalent to the "more.." link there is the important thing / what I was suggesting.

tl;dr: let users discover hotkeys, either by having ~ubiquitous "?" (which is enough of a standard around various sites at this point that most people might guess to look for it there), or a UI element on the page that teaches them about them, or both

**Comment by pkaminski** [comment]
Request for adding hotkeys to dashboard page noted.  In the meantime, there's just nothing to show in a cheatsheet there (and on other pages) unless I adopt GitHub's "global" cheatsheet approach.  I think context-sensitive cheatsheets are better, though.  Perhaps add a placeholder cheatsheet to pages without hotkeys that just says "No hotkeys available on this page" or some such?

As for a button to access the cheatsheet, I think that's overkill.  Only Asana does it AFAICT, and given the developer audience for Reviewable I'm comfortable assuming that folks know about "?".

**Comment by ryan-williams** [comment]
> I'm comfortable assuming that folks know about "?".

Except that "?" doesn't work on the homepage! :)

FWIW I also dislike github's context-sensitivity of what hotkeys it will show you when you press "?". Someone *using a hotkey* to open a modal to *learn more about hotkeys* is trying hard to learn how to be more of a poweruser, and we should let them! The state-dependent hiding feels like coddling in a bad way, in this context, to me. I'm open to a compromise where hotkeys not relevant to the current page are initially hidden in a "Show all" link or somesuch.

re: cheatsheet button being overkill, I disagree. You have a developer audience, and that means they are ppl that want hotkeys, **not** that they know the particular nuances of the implementation you've set up here; you've clearly done some good and powerful work around supporting this, but aspects of it are unique to reviewable, and I had no idea about it / have struggled to use it, as someone who really wants and likes hotkeys and have at least watched the discussion around them on reviewable from the early days via @danvk.

There's a giant "support" widget in the page chrome with its own menu, so there is plenty of room there or elsewhere on the page for this sort of thing.. just saying.

**Comment by ryan-williams** [comment]
also, https://reviewable.io/bindings.json is redirecting me to https://reviewable.io/ right now btw

**Comment by ryan-williams** [comment]
(linked from https://github.com/Reviewable/Reviewable/wiki/FAQ#can-i-customize-the-key-bindings )

**Comment by pkaminski** [comment]
Yeah, hence my idea about having "?" bring up a dialog on all pages, even ones without hotkeys.  I'll see if I can get that in.   I can also add a "Hotkeys" entry to the support menu, though the fact that there's a "custom key bindings URL" field in the account settings should be a pretty big tip-off too.  I'll need to think some more about the context-sensitivity of the cheatsheet (or lack thereof), but that has interesting interactions with customizability and isn't relevant anyway until we get commands implemented on other pages.

The missing `bindings.json` is a regression, I'll get that fixed ASAP.  In the meantime, here it is inlined so you're not blocked:

```
[
["n", "Next unreviewed file", "nextUnreviewedFile()"],
["p", "Previous unreviewed file", "prevUnreviewedFile()"],
[null, "Next file", "nextFile()"],
[null, "Previous file", "prevFile()"],
["shift+n", "Next changed file", "nextChangedFile()"],
["shift+p", "Previous changed file", "prevChangedFile()"],
["x", "Mark file as reviewed / unreviewed", "toggleCurrentFileReviewed()"],
["j", "Next unreplied comment", "nextUnrepliedDiscussion()"],
["k", "Previous unreplied comment", "prevUnrepliedDiscussion()"],
["?", "Show / hide keyboard shortcuts list", "toggleCheatSheet()"],
[null, "Next comment", "nextDiscussion()"],
[null, "Previous comment", "prevDiscussion()"],
["shift+j", "Next unresolved comment", "nextUnresolvedDiscussionWithoutDraft()"],
["shift+k", "Previous unresolved comment", "prevUnresolvedDiscussionWithoutDraft()"],
[null, "Next draft", "nextDraft()"],
[null, "Previous draft", "prevDraft()"],
[["r", "enter"], "Reply to comment", "replyToCurrentDiscussion()"],
["d", "Reply to comment with 'Done'", "replyToCurrentDiscussion('Done')"],
["y", "Acknowledge comment", "acknowledgeCurrentDiscussion()"],
["mod+shift+o", "Show / hide older comments", "toggleCurrentDiscussionOlderComments()"],
["mod+shift+p", "Toggle draft preview", "toggleCurrentDraftPreview()"]
]
```

**Comment by ryan-williams** [comment]
cool, thanks @pkaminski !

**Comment by pkaminski** [comment]
https://reviewable.io/bindings.json now works.  Sorry about that!

**Comment by pkaminski** [comment]
The keyboard shortcuts dialog is now available from every page, and has an entry in the Support menu.  Also, shortcuts are now mentioned in the tooltips (aka Guide), including any custom bindings.  And as a bonus I added some new actions to `bindings.json` (unbound by default) for people who really like the keyboard.  :smile:
